io-engine/src/grpc/v1/replica.rs

```diff
@@ -268,70 +290,83 @@ impl ReplicaRpc for ReplicaService {
         &self,
         request: Request<DestroyReplicaRequest>,
     ) -> GrpcResult<()> {
-        self.locked(GrpcClientContext::new(&request, function_name!()), async {
-            let args = request.into_inner();
-            info!("{:?}", args);
-            let rx = rpc_submit::<_, _, LvsError>(async move {
-                // todo: is there still a race here, can the pool be exported
-                //   right after the check here and before we
-                //   probe for the replica?
-                let lvs = match &args.pool {
-                    Some(destroy_replica_request::Pool::PoolUuid(uuid)) => {
-                        Lvs::lookup_by_uuid(uuid)
-                            .ok_or(LvsError::RepDestroy {
-                                source: Errno::ENOMEDIUM,
-                                name: args.uuid.to_owned(),
-                                msg: format!("Pool uuid={uuid} is not loaded"),
-                            })
-                            .map(Some)
-                    }
-                    Some(destroy_replica_request::Pool::PoolName(name)) => {
-                        Lvs::lookup(name)
-                            .ok_or(LvsError::RepDestroy {
-                                source: Errno::ENOMEDIUM,
-                                name: args.uuid.to_owned(),
-                                msg: format!("Pool name={name} is not loaded"),
-                            })
-                            .map(Some)
-                    }
-                    None => {
-                        // back-compat, we keep existing behaviour.
-                        Ok(None)
-                    }
-                }?;
-
-                let lvol = Bdev::lookup_by_uuid_str(&args.uuid)
-                    .and_then(|b| Lvol::try_from(b).ok())
-                    .ok_or(LvsError::RepDestroy {
-                        source: Errno::ENOENT,
-                        name: args.uuid.to_owned(),
-                        msg: "Replica not found".into(),
-                    })?;
-
-                if let Some(lvs) = lvs {
-                    if lvs.name() != lvol.pool_name()
-                        || lvs.uuid() != lvol.pool_uuid()
-                    {
-                        let msg = format!(
+        let ctx = GrpcClientContext::new(&request, function_name!());
+        let args = request.into_inner();
+        self.serialize_grpc(
+            ctx,
+            Some(args.clone().uuid),
+            self.init_subsystem().clone(),
+            true,
+            true,
```

Comment:
So we're making the replica create a global write operation?
Is it not enough to just lock the pool which this replica references?

Reply:
Changed the code, made import pool and create replica under same resource lock (`pool_name`) in place of global lock.